client/asset/eth/contractor.go

```diff
@@ -113,10 +113,19 @@ func newV0Contractor(net dex.Network, acctAddr common.Address, cb bind.ContractB
 }
 
 // initiate sends the initiations to the swap contract's initiate function.
-func (c *contractorV0) initiate(txOpts *bind.TransactOpts, contracts []*asset.Contract) (*types.Transaction, error) {
+func (c *contractorV0) initiate(txOpts *bind.TransactOpts, contracts []*asset.Contract) (tx *types.Transaction, err error) {
 	inits := make([]swapv0.ETHSwapInitiation, 0, len(contracts))
 	secrets := make(map[[32]byte]bool, len(contracts))
 
+	// If initiation fails with a provider, trust the next nonce.
+	if mRPC, is := c.cb.(*multiRPCClient); is {
+		defer func() {
+			if err != nil {
+				mRPC.voidUnusedNonce()
+			}
+		}()
+	}
+
```

Comment:
Do you mean confirm the tx was successful, get a conf?

Reply:
The wording "trust the next nonce" throws me.  Does this mean "allow the nonce in `txOpts` to be provided to `nextNonce` callers again"?  Because there are expected to be concurrent contractor calls, an `voidUnusedNonce` does not "release" the nonce for _this_ failed initiate transaction; it releases whatever is the **current** next nonce, which could be higher than this `txOpt`'s nonce... thus releasing the nonce that's used in a _different_ concurrent transaction that might succeed or have just succeeded already.

Reply:
OK so I guess we're preventing concurrent contractor calls via `w.nonceSendMtx`. I remember this now.  e.g.

```go
w.nonceSendMtx.Lock()
defer w.nonceSendMtx.Unlock()
txOpts, _ := w.node.txOpts(ctx, val, gasLimit, dexeth.GweiToWei(maxFeeRate))
return tx, w.withContractor(contractVer, func(c contractor) error {
	tx, err = c.initiate(txOpts, contracts)
	return err
})
```

So the idea is in `txOpts()` we grab the nonce, and if `initiate` fails, throw the nonce back into the pool.  I think that makes sense.

Reply:
Yes, throw the nonce back to the pool if we are pretty sure it can be used again. Then we still ask the provider for the next nonce anyway, which may be the same number or different.

Moving the error check up so it catches everything in the next iteration.

Reply:
> which could be higher than this txOpt's nonce... thus releasing the nonce that's used in a different concurrent transaction.

So, we do the concurrent calls, and the provider may give us the same nonce if they are close together. I guess it does not increase the nonce unless it sees a tx in mempool? I say this because even using the same provider sometimes gives us the same nonce, which brought up the need for the `*multiRPCClient.lastNonce` or something like it to "pause" the action and possibly fail if we are fairly sure we are reusing a nonce.

Reply:
Yeah definitely, the providers cannot be relieved relied upon for this .

💯